### PR TITLE
Document modes and subsystems

### PR DESCRIPTION
**Pull request by aartaka**
# Description

This documents lots of modes as an attempt to document everything for #2303 (with minor indentation and function signature fixes I was unable to resist). @Ambrevar, @aadcg, you can push your documentation improvements to this branch, if you want to.

**Manual testing and bugfixes don't belong to this PR, it's about documentation exclusively. Push to master or open separate PRs for that.**

# Discussion

## Documentation formats
It the course of documenting things, I've kind of settled at a certain structure for package and mode documentation. Packages are for implementation details and APIs, modes are for commands, configuration, and other user-facing things. That's why their documentation is interlinked, but different. Here's what I've arrived at:

### Package 
```
Package for `...-mode', mode for [description].

Implementation details (functions, macros, classes, sources etc.):
- [...]
- [...]
[...]

See the `...-mode' for the external user-facing APIs.
```

## #Mode
```
[Concise and plain language description of the mode]
[Additional details about it]

Commands:
[...]

Configuration options:
[...]

Example configuration:
[...]

See `nyxt/mode/...' package documentation for implementation details and internal programming APIs.
```

Obviously, these formats are merely a recommendation, because every mode is unique. But still, would be nice to have a certain consistency and structure. So @Ambrevar, @aadcg, try using these formats and feel free to suggest better ones for us to have!

## Responsibility distribution

@Ambrevar, can I take care of `repeat-mode`, `editor-mode`, and `cruise-control-mode` documentation now that they're broken/underdocumented?

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.

- [X] Git hygiene:
  - I have pulled from master before submitting this PR
  - There are no merge conflicts.
- [X] I've added the new dependencies as:
  - ASDF dependencies,
  - Git submodules,
    ```sh
	cd /path/to/nyxt/checkout
    git submodule add https://gitlab.common-lisp.net/nyxt/py-configparser _build/py-configparser
    ```
  - and Guix dependencies.
- [X] My code follows the style guidelines for Common Lisp code. See:
  - [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf)
  - [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml)
- [X] I have performed a self-review of my own code.
- [ ] My code has been reviewed by at least one peer.  (The peer review to approve a PR counts.  The reviewer must download and test the code.)
- [X] Documentation:
  - All my code has docstrings and `:documentation`s written in the aforementioned style.  (It's OK to skip the docstring for really trivial parts.)
  - I have updated the existing documentation to match my changes.
  - I have commented my code in hard-to-understand areas.
  - I have updated the `changelog.lisp` with my changes if it's anything user-facing (new features, important bug fix, compatibility breakage).
  - I have added a `migration.lisp` entry for all compatibility-breaking changes.
  - (If this changes something about the features showcased on Nyxt website) I have these changes described in the new/existing article at Nyxt website or will notify one of maintainters to do so.
- [X] Compilation and tests:
  - My changes generate no new warnings.
  - I have added tests that prove my fix is effective or that my feature works.  (If possible.)
  - I ran the tests locally (`(asdf:test-system :nyxt)` and `(asdf:test-system :nyxt/gi-gtk)`) and they pass.

### CODE REVIEW

**Comment by aadcg** [comment]
I'm reviewing this PR tomorrow, first thing in the morning. Thanks!

**Comment by aadcg** [comment]
With respect to the documentation format, any format is better than no format! By the principle of "monkey see, monkey do", future modes will follow that structure and that's positive indeed.

Remark: the discussion of the WHAT belong to Issues, not PRs. In PRs, you discuss the HOW. As you can imagine, no one will be unreasonable to the point of starting a discussion about the doc structure since, in the end, it could trigger re-doing a big portion of your work. In short, you're presenting us to that format post factum. 

---- 

I'll be pushing commits to this PR. You can revert them later as you please.

**Comment by Ambrevar** [comment]
> @Ambrevar, can I take care of `repeat-mode`, `editor-mode`, and `cruise-control-mode` documentation now that they're broken/underdocumented?

Please do :)

**Comment by Ambrevar** [comment]
I'm actually not too much in favour of command and option listing: they are cumbersome to maintain and risk running out of sync.

After all, package documentations should provide the user with the list of commands.  Same for class documentation and the slots.

**Comment by aadcg** [comment]
> I'm actually not too much in favour of command and option listing: they are cumbersome to maintain and risk running out of sync.

> After all, package documentations should provide the user with the list of commands. Same for class documentation and the slots.

That's exactly what I wanted to say, but [settled on a more polite observation](https://github.com/atlas-engineer/nyxt/pull/2957#issuecomment-1543484235).

We have `describe-mode`, and I think that's where we should fit the keybindings and user options (mode slots).

File: source/mode/history.lisp [hunk 1]
```diff
@@ -2,11 +2,47 @@
 ;;;; SPDX-License-Identifier: BSD-3-Clause
 
 (nyxt:define-package :nyxt/mode/history
-    (:documentation "Mode to store current buffer navigation into the global history."))
+  (:documentation "Mode to persist navigation into the global history and navigate around it.
```
Comment (aadcg) [inline source/mode/history.lisp:5]: Try to put yourself in the shoes of a newcomer. A mode that persists navigation and enabling navigating in this thing called "global history". Hmm... 

Now let's look at the old version.

> Mode to store current buffer navigation into the global history.

Well, I can actually imagine what it does!

File: source/mode/keyscheme.lisp [hunk 2]
```diff
@@ -3,12 +3,27 @@
 
 (nyxt:define-package :nyxt/mode/keyscheme
   (:documentation "All modes that set `keyscheme' should inherit from this mode.
-Ensures that a single keybindings mode, such as `nyxt/mode/emacs', is enabled."))
+Ensures that a single keybindings mode, such as
+`nyxt/mode/emacs'/`keymaps:emacs', is enabled."))
```
Comment (aadcg) [inline source/mode/keyscheme.lisp:7]: Honestly, it was much better as it was. I don't understand the meaning of:

> Ensures that a single keybindings mode, such as `nyxt/mode/emacs'/`keymaps:emacs', is enabled.

But I think it means "Ensures that a single keyscheme is enables", which is already clear from "All modes that set `keyscheme' should inherit from this mode.".

File: source/mode/keyscheme.lisp [hunk 3]
```diff
@@ -3,12 +3,27 @@
...
 (in-package :nyxt/mode/keyscheme)
 
 (define-mode keyscheme-mode ()
   "All modes that set `keyscheme' should inherit from this mode.
-Ensures that a single keybindings mode, such as `nyxt/mode/emacs', is enabled."
+Ensures that a single keybindings mode, such as
+`nyxt/mode/emacs'/`keymaps:emacs', is enabled.
```
Comment (aadcg) [inline source/mode/keyscheme.lisp:13]: Same.

File: source/mode/no-procrastinate.lisp [hunk 4]
```diff
@@ -3,12 +3,25 @@
...
+entries. `no-procrastinate-add' is the wrapper around it to add new entries.
+
+`serialize-object' is the method to print these hosts to the Lisp file
+`read'ably.
+
+The mode itself inherits from `nyxt/mode/blocker:blocker-mode' and reuses its APIs."))
 (in-package :nyxt/mode/no-procrastinate)
 
+;;; TODO: Remove? It's mostly poorly duplicating the blocker-mode logic and is
+;;; poorly written, rotting in the dark.
```
Comment (aadcg) [inline source/mode/no-procrastinate.lisp:18]: This comment  begs the question "is the effort put in documenting the mode worth it"?

I will not review this commit.

**Review by aadcg — CHANGES_REQUESTED** [review]
@aartaka I've left some comments in the PR's diff, sent some commits and made some general observations.

@Ambrevar's observation about the documentation structure makes sense to me.

File: source/mode/no-procrastinate.lisp (hunk 4 above)
Comment (aartaka) [reply source/mode/no-procrastinate.lisp:18]: It kinda isn't worth it, yes. But still, for completeness sake...

File: source/mode/keyscheme.lisp (hunk 2 above)
Comment (aartaka) [reply source/mode/keyscheme.lisp]: Done in 5cc8709e37ed8fc035c397cdb22647593f21f1bd!

File: source/mode/history.lisp (hunk 1 above)
Comment (aartaka) [reply source/mode/history.lisp]: Reverting in cb6916944c9c90cf740a226776a8cefd51002494 ೖ(⑅σ̑ᴗσ̑)ೖ

**Comment by aartaka** [comment]
Indeed, @Ambrevar, you're correct. Especially given that mode `describe-*` page now lists all the commands belonging to the mode （；¬＿¬)

Command listings in mode docs are removed in 23ceaac64969d9d35dc65ce03c0738baf505772d (but I've retained package documentation with internals, because it's useful for hacking on said packages.)

File: source/mode/repeat.lisp [hunk 5]
```diff
@@ -58,38 +81,39 @@ It takes a `repeat-mode' instance as argument.")
...
+  (enable-modes* 'repeat-mode
+                 (current-buffer)
+                 :repeat-count times
+                 :repeat-action function))
 
 (defvar *repeat-times-stack* 0
   "The current number of repetitions.")
 
 (defun make-repeat-command-dispatcher (times)
+  "Create a command dispatcher that counts the M-digic keys and adds them together.
```
Comment (aadcg) [inline source/mode/repeat.lisp:114]: I missed a typo here! "digit" instead of "digic".

**Review by aadcg — APPROVED** [review]

**Comment by aadcg** [comment]
Sure! I forgot about it. 

I agree about squashing it all down to a single commit.

**Comment by aartaka** [comment]
Pushed with bd709c9c21aa4c4c944bbade427d775d02160294!

**Comment by aadcg** [comment]
Thanks!

**Comment by aartaka** [comment]
@aadcg, there are several modes that you're to document, if you don't mind (I've chaotically documented modes from `annotate` to `cruise-control', sorry):
```
document
download
emacs
expedition
file-manager
force-https
```

And then @Ambrevar:
```
process
prompt-buffer
proxy
reading-line
record-input-field
reduce-bandwidth
reduce-tracking
remembrance
repeat
repl
search-buffer
small-web
spell-check
style
tts
user-script
vi
visual
watch
```

**Pull request merged**